3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,253 @@
+import {AMP_ANALYTICS_3P_MESSAGE_TYPE} from '../src/3p-analytics-common';
+initLogConstructor();
+// TODO(alanorozco): Refactor src/error.reportError so it does not contain big
+// transitive dependencies and can be included here.
+setReportError(() => {});
+
+/** @private @const {string} */
+const TAG_ = 'ampanalytics-lib';
+
+/**
+ * Receives messages bound for this cross-domain iframe, from all creatives
+ */
+class AmpAnalytics3pMessageRouter {
+  /** @param {!Window} win */
+  constructor(win) {
+
+    /** @private {!Window} */
+    this.win_ = win;
+
+    try {
+      const frameNameData = JSON.parse(this.win_.name);
+      if (!frameNameData.sentinel) {
+        const ex = {message: 'JSON parsed, but did not include sentinel.'};
+        throw ex;
+      }
+      // The sentinel is required by iframe-messaging-client, and is used to
+      // uniquely identify the frame as part of message routing
+      /** @const {string} */
+      this.sentinel_ = frameNameData.sentinel;
+    } catch (e) {
+      dev().warn(TAG_, 'Unable to set sentinel from name attr: ' +
+        this.win_.name + ': ' + e.message);
+      return;
+    }
+
+    /** @private {!Object<string, !AmpAnalytics3pCreativeMessageRouter>} */
+    this.creativeMessageRouters_ = {};
+
+    /**
+     * Handles communication between frames
+     * @private {!IframeMessagingClient}
+     */
+    this.iframeMessagingClient_ = new IframeMessagingClient(win);
+    this.iframeMessagingClient_.setSentinel(this.sentinel_);
+    this.iframeMessagingClient_.registerCallback(
+      AMP_ANALYTICS_3P_MESSAGE_TYPE.CREATIVES,
+      envelope => {
+        dev().assert(envelope[AMP_ANALYTICS_3P_MESSAGE_TYPE.CREATIVES],
+          'Received empty events envelope');
+        // Iterate through the array of "new creative" events, and for each
+        // one, create an AmpAnalytics3pCreativeMessageRouter.
+        envelope[AMP_ANALYTICS_3P_MESSAGE_TYPE.CREATIVES].forEach(creative => {
+          this.creativeMessageRouters_[creative.senderId] =
+            this.creativeMessageRouters_[creative.senderId] ||
+            new AmpAnalytics3pCreativeMessageRouter(this, this.sentinel_,
+              creative.senderId);
+          this.creativeMessageRouters_[creative.senderId]
+            .receiveNewCreativeMessage(creative);
+        });
+      });
+    this.iframeMessagingClient_.registerCallback(
+      AMP_ANALYTICS_3P_MESSAGE_TYPE.EVENTS,
+      envelope => {
+        dev().assert(envelope[AMP_ANALYTICS_3P_MESSAGE_TYPE.EVENTS],
+          'Received empty events envelope');
+        // Iterate through the array of events, dispatching each to the
+        // listener for the appropriate creative
+        envelope[AMP_ANALYTICS_3P_MESSAGE_TYPE.EVENTS].forEach(event => {
+          dev().assert(this.creativeMessageRouters_[event.senderId],
+            'Received event message prior to new creative message.' +
+            ' Discarding.');
+          this.creativeMessageRouters_[event.senderId]
```

Comment:
Perhaps the format of the message should be senderId to its events?  This would greatly reduce the iterations and allow you to pass an array of events which I think makes it clearer to the vendor that events are batched.